ivm/src/ext.rs

```diff
@@ -103,25 +140,42 @@ impl<'ivm> ExtVal<'ivm> {
 
   /// Creates a new `ExtVal` with a given type and payload.
   #[inline(always)]
-  pub const fn new(ty: ExtTy<'ivm>, payload: u32) -> Self {
-    Self(payload as (u64) << 32 | ty.id() as (u64) << 16 | Tag::ExtVal as u64, PhantomData)
+  pub fn new(ty_id: ExtTyId<'ivm>, payload: u64) -> Self {
+    debug_assert!(
+      payload & !(Self::PAYLOAD_MASK >> 3) == 0,
+      "ExtVal::new with non-payload bits set"
+    );
+
+    unsafe {
+      Self::from_bits(
+        ((ty_id.bits() as u64) << 48)
+          | ((payload << 3) & Self::PAYLOAD_MASK)
+          | (Tag::ExtVal as u64),
+      )
+    }
   }
-  /// Accesses the type of this value.
+
+  /// Returns the type id of this value.
   #[inline(always)]
-  pub fn ty(&self) -> ExtTy<'ivm> {
-    ExtTy((self.0 >> 16) as u16, self.1)
+  pub fn ty_id(&self) -> ExtTyId<'ivm> {
+    ExtTyId::from_bits((self.0 >> 48) as u16)
   }
 
-  /// Accesses the payload of this value.
+  /// Returns the 45-bit payload of this value as the _lowest_ 45 bits in a
+  /// `u64`. The highest 19 bits are guaranteed to be zero.
   #[inline(always)]
-  pub fn payload(&self) -> u32 {
-    (self.0 >> 32) as u32
+  fn payload(&self) -> u64 {
+    (self.0 & Self::PAYLOAD_MASK) >> 3
```

Comment:
Return an unshifted payload, where top 16-bits and bottom 3 bits are guaranteed to be zero.